siliconcompiler/schema/schema_obj.py

```diff
@@ -194,204 +229,162 @@ def record_history(self):
+
+        The expected type is based on the schema parameter type string for
+        value-related fields, and is based on the field itself for other fields.
+        This function raises a TypeError if an illegal value is provided.
+
+        The normalization process provides some leeway in how users supply
+        values, while ensuring that values are stored consistently in the schema.
+
+        The normalization rules are as follows:
+        - If a scalar is provided for a list type, it is promoted to a list of
+        one element.
+        - If a list is provided for a tuple type, it is cast to a tuple (since
+        the JSON module serializes tuples as arrays, which are deserialized into
+        lists).
+        - Elements inside lists and tuples are normalized recursively.
+        - All non-list values have a string representation that gets cast to a
+        native Python type (since we receive strings from the CLI):
+          - bool: accepts "true" or "false"
+          - ints and floats: cast as if by int() or float()
+          - tuples: accepts comma-separated values surrounded by parens
         '''
-        Internal recursive function that searches the Chip schema for a
-        match to the combination of *args and fields supplied. The function is
-        used to set and get data within the dictionary.
 
-        Args:
-            cfg(dict): The cfg schema to search
-            keypath (str): Concatenated keypath used for error logging.
-            args (str): Keypath/value variable list used for access
-            field(str): Leaf cell field to access.
-            mode(str): Action (set/get/add/getkeys/getkeys)
-            clobber(bool): Specifies to clobber (for set action)
+        if value is None and not Schema._is_list(field, sc_type):
+            # None is legal for all scalars, but not within collection types
+            # TODO: could consider normalizing "None" for lists to empty list?
+            return value
+
+        if field in ('value', 'defvalue'):
+            try:
+                return Schema._normalize_value(value, sc_type)
+            except TypeError:
+                # Re-raise exception with a consistent message
+                raise TypeError(f'Invalid value {value}: expected {sc_type}') from None
+        else:
+            return Schema._normalize_field(value, sc_type, field)
 
-        '''
+    @staticmethod
+    def _normalize_value(value, sc_type):
+        if sc_type.startswith('['):
+            if not isinstance(value, list):
+                value = [value]
+            base_type = sc_type[1:-1]
+            return [Schema._normalize_value(v, base_type) for v in value]
+
+        if sc_type.startswith('('):
+            # TODO: make parsing more robust to support tuples-of-tuples
+            if isinstance(value, str):
+                value = value[1:-1].split(',')
+            elif not (isinstance(value, tuple) or isinstance(value, list)):
+                raise TypeError
```

Comment:
These `TypeErrors` are caught by the user of this function and re-raised with a message (I noticed I was copy-pasting the same message for each). I'm happy to add messages back to this method to make it more explicit what's going on, though.

Reply:
Can we support our standard logger messages for this? What does the TypeError report back?  Some of the error messages are not exactly user friendly. We should be able to help the user quickly figure out what they did wrong with the right fields (value, sc_type, keypath?

Reply:
We can't support our standard logger messages directly, since we're not in the Chip object. I've just opened an issue tracking this, so we can fix it eventually: #1231. 

However, note that the TypeErrors raised by these functions are caught and have their messages re-routed to the normal error logging mechanism inside the equivalent Chip object methods.

To make the behavior more explicit, I've added messages to all of these bare TypeErrors, using a variable to avoid repeating the message in multiple places.